File: 01_GOF23/DP_examples/Lua/02_Structural/Flyweight/UnsharedFlyweight.lua
-- 加载 Flyweight 模块
require("Flyweight")

-- 非共享具体享元类 - 圆形增强
---@class CircleAddition : IShape
---@field shape Circle
CircleAddition = IShape:new()
---@param shape Circle
function CircleAddition:new(shape)
    local o = IShape.new(self)
    o.shape = shape
    return o
end
function CircleAddition:draw()
    self.shape:draw()
end
function CircleAddition:fill()
    self.shape:fill()
end

-- 移动位置
function CircleAddition:movePosition(x, y)
    print(string.format("Move Circle to (%.2f, %.2f)", x, y))
end

-- 重置半径
function CircleAddition:resetRadius(radius)
    print(string.format("Reset Radius to %.2f", radius))
end

File: 01_GOF23/DP_examples/Rust/02_Structural/Flyweight/unshared_flyweight.rs
use crate::component::{IShape, Circle, CircleData};

// 非共享具体享元类 - 圆形增强
pub struct CircleAddition {
    shape: Box<dyn IShape>,
}

impl CircleAddition {
    pub fn new(shape: Box<dyn IShape>) -> Self {
        Self {
            shape,
        }
    }
    
    pub fn move_position(&self, x: f32, y: f32) {
        println!("Move Circle to ({}, {})", x, y);
    }
    
    pub fn reset_radius(&self, radius: f32) {
        println!("Reset Radius to {}", radius);
    }
}

impl IShape for CircleAddition {
    fn draw(&self) {
        self.shape.draw();
    }
    
    fn fill(&self) {
        self.shape.fill();
    }
}

// 圆形工厂
pub struct CircleFactory;

impl CircleFactory {
    pub fn new() -> Self {
        Self
    }
    
    pub fn get_circle(&self, color: &str, radius: f32, x: f32, y: f32) -> Circle {
        let data = CircleData {
            color: color.to_string(),
            radius,
        };
        Circle::new(data, x, y)
    }
    
    pub fn get_circle_by_data(&self, data: &CircleData, x: f32, y: f32) -> Circle {
        Circle::new(data.clone(), x, y)
    }
    
    pub fn get_spacial_circle(&self, circle: Circle) -> CircleAddition {
        CircleAddition::new(Box::new(circle))
    }
}
